packages/core/test/features/st-mixin.spec.ts

```diff
@@ -409,37 +406,174 @@ describe(`features/st-mixin`, () => {
+                    }
+                }
+            `)
+            );
+        });
+        it('should preserve nesting selector', () => {
+            const { sheets } = testStylableCore(`
+                .mix {
+                    &.inner {
+                        color: green;
+                    }
+                }
+
+                .into {
+                    -st-mixin: mix;
+                }
+            `);
+
+            const { meta } = sheets['/entry.st.css'];
+
+            expect(deindent(meta.targetAst!.toString())).to.eql(
+                deindent(`
+                .entry__mix {
+                    &.entry__inner {
+                        color: green;
+                    }
+                }
+
+                .entry__into {
+                    &.entry__inner {
+                        color: green;
+                    }
+                }
+            `)
+            );
+        });
+        it('should mix into nesting context', () => {
+            const { sheets } = testStylableCore(`
+                .mix {
+                    &.inner {
+                        color: green;
+                    }
+                }
+
+                .intoTop {
+                    &.intoNested {
+                        -st-mixin: mix;
+                    }
+                }
+            `);
+
+            const { meta } = sheets['/entry.st.css'];
+
+            expect(deindent(meta.targetAst!.toString())).to.eql(
+                deindent(`
+                .entry__mix {
+                    &.entry__inner {
+                        color: green;
+                    }
+                }
+
+                .entry__intoTop {
+                    &.entry__intoNested {
+                    &.entry__inner {
+                        color: green;
+                    }
+                    }
+                }
+            `)
+            );
+        });
+        it('should report recursive mixin ', () => {
+            const { sheets } = testStylableCore(`
+                .mix {
+                    &.inner.mix {
+                        color: green;
+                    }
+                }
+
+                /* @transform-error ${mixinDiagnostics.INVALID_RECURSIVE_MIXIN()}*/
+                .into {
+                    -st-mixin: mix;
+                }
+            `);
+
+            const { meta } = sheets['/entry.st.css'];
+
+            expect(deindent(meta.targetAst!.toString())).to.eql(
+                deindent(`
+                .entry__mix {
+                    &.entry__inner.entry__mix {
+                        color: green;
+                    }
+                }
+
+                /* @transform-error ${mixinDiagnostics.INVALID_RECURSIVE_MIXIN()}*/
+                .entry__into {
+                    &&.entry__inner {
+                        color: green;
+                    }
+                }
+            `)
+            );
+        });
+        it('should mix nested at-rules with content', () => {
```

Comment:
we should test the case when there are decls after the -st-mixin.

Reply:
check for what? that we don't mix nested rules before declarations? split the rule in that case?

Reply:
either that or not allowing

Reply:
went with nesting in case of declaration and nested rules out of order